### PR TITLE
Adaround Implementation

### PR DESCRIPTION
**Pull request by edmundw314**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41881 Adaround Implementation**

Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

### CODE REVIEW

File: test/quantization/test_adaround.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import torch
+import torch.nn as nn
+from torch.testing._internal.common_quantization import QuantizationTestCase
+from torch.quantization import QuantStub, DeQuantStub
+from torch.quantization._adaround import _supported_modules
```
Comment (jerryzh168) [inline test/quantization/test_adaround.py:5]: lint

File: torch/quantization/_adaround.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+def parent_child_names(name):
+    '''Splits full name of submodule into parent submodule's full name and submodule's name
+    '''
+    split_name = name.rsplit('.', 1)
+    if len(split_name) == 1:
+        return '', split_name[0]
+    else:
+        return split_name[0], split_name[1]
+
+class adaround(FakeQuantize):
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:45]: nit: AdaRound for class name

Comment (jerryzh168) [reply torch/quantization/_adaround.py:45]: maybe `AdaRoundFakeQuantize` to make it more clear that this is a fake quantize module

File: torch/quantization/_adaround.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    def __init__(self, *args, **keywords):
+        super(adaround, self).__init__(*args, **keywords)
+        self.continous_V = None
+
+    def forward(self, X):
+        if self.continous_V is None:
+            # small initial values are intended so no bias is introduced on initialization or calibration
+            self.continous_V = torch.nn.Parameter(torch.ones(X.size()) / 10000)
+
+        if self.observer_enabled[0] == 1:
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:55]: can you add a comment on why we are not reusing the forward of parent module? I already forgot what reason is actually

File: torch/quantization/_adaround.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+            if self.qscheme == torch.per_channel_symmetric or self.qscheme == torch.per_channel_affine:
+                X = torch.fake_quantize_per_channel_affine(X, self.scale, self.zero_point,
+                                                           self.ch_axis, self.quant_min, self.quant_max)
+            else:
+                X = torch.fake_quantize_per_tensor_affine(X, float(self.scale),
+                                                          int(self.zero_point), self.quant_min,
+                                                          self.quant_max)
+        return X
+
+    def randomize(self):
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:73]: feel free to remove if this is not used

File: torch/quantization/_adaround.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+        self.wrapped_module.activation_post_process.enable_fake_quant()
+        self.wrapped_module.weight_fake_quant.enable_fake_quant()
+        self.quantized_output = self.wrapped_module(x)
+
+        print("norm in forward outputwrapper: ", torch.norm(self.float_output - self.quantized_output))
+        return self.dequant(self.quantized_output)
+
+
+araround_fake_quant = adaround.with_args(observer=HistogramObserver, quant_min=-128, quant_max=127,
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:103]: default_adaround_fake_quant?

Comment (jerryzh168) [inline torch/quantization/_adaround.py:103]: btw, using HistogramObserver might be the reason that it is slow, you can try MinMax to make it faster when you run tests.

File: torch/quantization/_adaround.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+from torch.quantization.qconfig import *
+from torch.quantization.fake_quantize import *
+import torch.nn.qat.modules as nnqat
+from torch.quantization.default_mappings import DEFAULT_QAT_MODULE_MAPPING
+from torch.quantization import QuantStub, DeQuantStub
+import copy
+_supported_modules = {nn.Conv2d, nn.Linear}
+
+# Hyper parameters for loss function
+beta_high = 8
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:16]: for constant we typically use all capitable case, e.g. BETA_HIGH = 8

Comment (jerryzh168) [reply torch/quantization/_adaround.py:16]: also some of these parameters should probably be configurable in the adaround module I think

File: torch/quantization/_adaround.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+beta_low = 2
+norm_scaling = 10
+regularization_scaling = .1
+
+# Adaround training parameters
+number_of_epochs = 10
+number_of_calibration_batches = 10
+learning_rate = .1
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:24]: these should probably be argument for learn_adaround function

File: torch/quantization/_adaround.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/_adaround.py:45]: also this should probably be in a different file

File: torch/quantization/_adaround.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+        if self.observer_enabled[0] == 1:
+            self.activation_post_process(X.detach())
+            _scale, _zero_point = self.calculate_qparams()
+            _scale, _zero_point = _scale.to(self.scale.device), _zero_point.to(self.zero_point.device)
+            self.scale = _scale
+            self.zero_point = _zero_point
+
+        if self.fake_quant_enabled[0] == 1:
+            X = adaround_rounding(self, X)
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:63]: why is this not a method of the adaround fake quantize module?

File: torch/quantization/_adaround.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    # print("sqnr of float and quantized: ", computeSqnr(float_weight, quantized_weight))
+    return Frobenius_norm * norm_scaling + regularization_scaling * regulization
+
+def loss_function(model, count):
+    ''' Given model, the loss function of all of its whitelisted leaf modules will
+    be added up and returned.
+    '''
+    result = torch.Tensor([0])
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, OuputWrapper):
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:171]: do we need this wrapper? can you just check the type of the module directly? if it is AdaRoundFakeQuantize module then we will accumulate the loss

File: torch/quantization/_adaround.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    scale = model.scale
+
+    weights_divided_by_scale = torch.div(weight, scale)
+    weights_divided_by_scale = torch.floor(weights_divided_by_scale)
+    weights_clipped = weights_divided_by_scale + clipped_sigmoid(continous_V)
+
+    weights_w_adaround_rounding = scale * torch.clamp(weights_clipped, model.quant_min, model.quant_max)
+    return weights_w_adaround_rounding
+
+def loss_function_leaf(model, count):
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:135]: this can probably also be a part of the AdaRoundFakeQuantize module

File: torch/quantization/_adaround.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+def loss_function(model, count):
+    ''' Given model, the loss function of all of its whitelisted leaf modules will
+    be added up and returned.
+    '''
+    result = torch.Tensor([0])
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, OuputWrapper):
+            result = result + loss_function_leaf(submodule, count)
+            result = result * .90
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:173]: what is this?

File: torch/quantization/_adaround.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+    return result
+
+def add_wrapper_class(model, white_list=DEFAULT_QAT_MODULE_MAPPING.keys()):
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:176]: nit: white_list --> supported_list

File: torch/quantization/_adaround.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+    '''
+    for name, submodule in model.named_modules():
+        if type(submodule) in white_list:
+            parent_name, child_name = parent_child_names(name)
+            parent_module = get_module(model, parent_name)
+            parent_module._modules[child_name] = OuputWrapper(submodule)
+            submodule.qconfig = adaround_qconfig
+
+
+def prepare_adaround(float_model, dataset, white_list=_supported_modules):
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:188]: same here, change white_list

File: torch/quantization/_adaround.py [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+def add_wrapper_class(model, white_list=DEFAULT_QAT_MODULE_MAPPING.keys()):
+    ''' Throws on a wrapper class to collect the last output passed through it.
+    This information is used in computing the loss function.
+    '''
+    for name, submodule in model.named_modules():
+        if type(submodule) in white_list:
+            parent_name, child_name = parent_child_names(name)
+            parent_module = get_module(model, parent_name)
+            parent_module._modules[child_name] = OuputWrapper(submodule)
+            submodule.qconfig = adaround_qconfig
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:185]: these things should be configurable by user I think

File: torch/quantization/_adaround.py [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+                                                          int(self.zero_point), self.quant_min,
+                                                          self.quant_max)
+        return X
+
+    def randomize(self):
+        # uniform distribution of vals
+        init.kaiming_uniform_(self.continous_V, a=math.sqrt(5))
+
+
+class OuputWrapper(nn.Module):
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:78]: can we do this in the forward of adaround fake quantize module?

File: torch/quantization/_adaround.py (hunk 13 above)
Comment (jerryzh168) [inline torch/quantization/_adaround.py:188]: users are supposed to attach qconfig themselves in eager mode, I don't think we should call prepare_qat for user. AdaRoundFakeQuantize module will just be a module that user can use in replacement of FakeQuantize module. we should implement loss functions etc. as a method of AdaRoundFakeQuantize module itself.

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
I think in general we should move most of the logic to AdaRoundFakeQuantize module

File: test/quantization/test_adaround.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
+import torch
+import torch.nn as nn
+from torch.testing._internal.common_quantization import QuantizationTestCase
+from torch.quantization import QuantStub, DeQuantStub
+import torch.quantization._adaround as _adaround
+
+
+class TestAdaround(QuantizationTestCase):
+    def single_layer_adaround(self, model, adaround_func, img_data):
```
Comment (jerryzh168) [inline test/quantization/test_adaround.py:9]: this is not implemented yet

I think we can compare with default fake quant and check if the sqnr is always better

File: torch/quantization/adaround_fake_quantize.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+                                           'beta_high': 8,
+                                           'beta_low': 2,
+                                           'norm_scaling': 10,
+                                           'regularization_scaling': .1}
+
+    def __init__(self, *args, **keywords):
+        settings = AdaRoundFakeQuantize.adaround_default_attribute_settings
+
+        for attribute in settings:
+            if attribute in settings:
```
Comment (jerryzh168) [inline torch/quantization/adaround_fake_quantize.py:28]: you mean `if attribute in keywords`?

File: torch/quantization/adaround_fake_quantize.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+                setattr(self, attribute, keywords[attribute])
+                del keywords[attribute]
```
Comment (jerryzh168) [inline torch/quantization/adaround_fake_quantize.py:30]: why do we need to delete here?

File: torch/quantization/adaround_fake_quantize.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    quantization results.
+
+    '''
+    adaround_default_attribute_settings = {'continous_V': None,
+                                           'beta_high': 8,
+                                           'beta_low': 2,
+                                           'norm_scaling': 10,
+                                           'regularization_scaling': .1}
+
+    def __init__(self, *args, **keywords):
```
Comment (jerryzh168) [inline torch/quantization/adaround_fake_quantize.py:24]: I feel it might be better to just spell out the arguments in the list, also please add a comment for all the arguments

File: torch/quantization/adaround_fake_quantize.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                setattr(self, attribute, settings[attribute])
+
+        self.tuning = False
+        super(AdaRoundFakeQuantize, self).__init__(*args, **keywords)
+
+    def forward(self, X):
+        if self.continous_V is None:
+            # need some value already assigned to continuous_V for kaiming to work
+            self.continous_V = torch.nn.Parameter(torch.zeros(X.size()))
+            init.kaiming_uniform_(self.continous_V, a=math.sqrt(5))
```
Comment (jerryzh168) [inline torch/quantization/adaround_fake_quantize.py:41]: this should be done in __init__?

File: torch/quantization/adaround_fake_quantize.py [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+
+        if self.tuning:
```
Comment (jerryzh168) [inline torch/quantization/adaround_fake_quantize.py:43]: `tuning` --> `enable_adaround`

File: torch/quantization/adaround_fake_quantize.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+
+
+        clip_V = self.clipped_sigmoid(self.continous_V)
+        spreading_range = torch.abs((2 * clip_V) - 1)
+        one_minus_beta = 1 - (spreading_range ** beta)
+        regulization = torch.sum(one_minus_beta)
+
+        return Frobenius_norm * self.norm_scaling + self.regularization_scaling * regulization
+
+parameters = {'observer': MovingAverageMinMaxObserver, 'quant_min': -128, 'quant_max': 127, 'dtype': torch.qint8,
```
Comment (jerryzh168) [inline torch/quantization/adaround_fake_quantize.py:95]: these can be removed when you have default argument

File: torch/quantization/_adaround.py [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+        number_of_epochs: number of batches each layer is trained on
+        learning_rate: learning rate of the training
+    '''
+
+    def optimize_V(leaf_module):
+        '''Takes in a leaf module with an adaround attached to its
+        weight_fake_quant attribute and runs an adam optimizer on the continous_V
+        attribute on the adaround module
+        '''
+        leaf_module.weight_fake_quant.tuning = True
```
Comment (jerryzh168) [inline torch/quantization/_adaround.py:55]: is this needed?

**Review by jerryzh168 — COMMENTED** [review]
see inline comments

File: torch/quantization/adaround_fake_quantize.py (hunk 18 above)
Comment (edmundw314) [reply torch/quantization/adaround_fake_quantize.py:30]: keywords gets passed to its parent class FakeQuantize, i thought weird things would happen if im passing extra parameters to FakeQuantize

File: test/quantization/test_adaround.py (hunk 16 above)
Comment (raghuramank100) [reply test/quantization/test_adaround.py:9]: We should compare with default PTQ, not fake-quant.

File: torch/quantization/adaround_fake_quantize.py (hunk 20 above)
Comment (edmundw314) [reply torch/quantization/adaround_fake_quantize.py:41]: can't be done in init because you need to know X.size() (which is unknown when init is called), to get around that i set continous_V to None and the first time the forward is run we get X.size() and its updated accordingly

File: torch/quantization/_adaround.py (hunk 23 above)
Comment (edmundw314) [reply torch/quantization/_adaround.py:55]: weird things happened when i didn't separate the training for continous_V from the normal qat calibration, if we get rid of this the qat logic has trouble finding good values for scale and offset i think (the accuracy in demos plummets to ~15-17% when we don't add this)

File: torch/quantization/adaround_fake_quantize.py (hunk 20 above)
Comment (jerryzh168) [reply torch/quantization/adaround_fake_quantize.py:41]: oh, I see, please add a comment

File: torch/quantization/adaround_fake_quantize.py (hunk 18 above)
Comment (jerryzh168) [reply torch/quantization/adaround_fake_quantize.py:30]: you should call super first with explicit arguments and then initialize the remaining attributes

File: torch/quantization/_adaround.py (hunk 23 above)
Comment (jerryzh168) [reply torch/quantization/_adaround.py:55]: oh, are you using the same model to do the comparison?

File: test/quantization/test_adaround.py (hunk 16 above)
Comment (jerryzh168) [reply test/quantization/test_adaround.py:9]: how is default fake quant without learning weight different from PTQ?

Comment (jerryzh168) [reply test/quantization/test_adaround.py:9]: @edmundw314 pleases make sure when you compare with default fake quant do not make the weight learnable. it should just be observing and fake quant, but without backprop or the optimizer should not have any parameters.
